File: 1-intro/max-value.js
// Time complexity: O(n) linear
// Space complexity: O(1) constant
const maxValue = (nums) => {
    let max = nums[0];
    nums.forEach( number => {
        if(number > max) {
            max = number;
        }
    })
    return max;
};

// Time complexity: O(n^2) due to slice
// Space complexity: O(n) because recursion (memory stack)
const maxValueRecursive = (nums) => {
    if(!nums) return -Infinity;
    if(nums.length === 1) return nums[0];
    
    let max = maxValue(nums.slice(1));
    if(nums[0] > max) max = nums[0]
    return max;
};
  
// Time complexity: O(n) linear
// Space complexity: O(1) constant
const maxValue2 = (nums) => {
  let max = -Infinity;
  for(let num of nums) {
    if(num > max) max = num;
  }
  return max;
};

File: 0-intro/max-value.py
def max_value(nums):
  max = float('-inf')

  for num in nums:
    if (num > max):
      max = num

  return max

# print(max_value([4, 7, 2, 8, 10, 9]))
# print(max_value([10, 5, 40, 40.3])) # -> 40.3
# print(max_value([-5, -2, -1, -11])) # -> -1
# print(max_value([42])) # -> 42
# print(max_value([1000, 8])) # -> 1000
# print(max_value([1000, 8, 9000])) # -> 9000
# print(max_value([2, 5, 1, 1, 4])) # -> 5
